File: piwigo/Album/Cells/AlbumTableViewCell.m
//
//  AlbumTableViewCell.m
//  piwigo
//
//  Created by Spencer Baker on 1/24/15.
//  Copyright (c) 2015 bakercrew. All rights reserved.
//

#import "AlbumService.h"
#import "AlbumTableViewCell.h"
#import "CategoriesData.h"
#import "MBProgressHUD.h"
#import "NetworkHandler.h"
#import "PiwigoAlbumData.h"

NSString * const kAlbumTableCell_ID = @"AlbumTableViewCell";

@interface AlbumTableViewCell()

@end

@implementation AlbumTableViewCell

-(void)imageUpdated
{
    self.backgroundImage.image = self.albumData.categoryImage;
}

-(void)setupWithAlbumData:(PiwigoAlbumData*)albumData
{
    if(!albumData) return;
    
    self.albumData = albumData;
    
    // General settings
    self.backgroundColor = [UIColor piwigoColorBackground];
    self.contentView.layer.cornerRadius = 14;
    self.contentView.backgroundColor = [UIColor piwigoColorCellBackground];
    self.selectionStyle = UITableViewCellSelectionStyleNone;
    self.topCut.layer.cornerRadius = 7;
    self.topCut.backgroundColor = [UIColor piwigoColorBackground];
    self.bottomCut.layer.cornerRadius = 7;
    self.bottomCut.backgroundColor = [UIColor piwigoColorBackground];

    // Album name
    self.albumName.text = self.albumData.name;
    self.albumName.font = [UIFont piwigoFontButton];
    self.albumName.textColor = [UIColor piwigoColorOrange];
    self.albumName.font = [self.albumName.font fontWithSize:[UIFont fontSizeForLabel:self.albumName nberLines:2]];

    // Album comment
    if (self.albumData.comment.length == 0) {
        if(NetworkVarsObjc.hasAdminRights) {
            self.albumComment.text = [NSString stringWithFormat:@"(%@)", NSLocalizedString(@"createNewAlbumDescription_noDescription", @"no description")];
            self.albumComment.textColor = [UIColor piwigoColorRightLabel];
        } else {
            self.albumComment.text = @"";
        }
    }
    else {
        self.albumComment.text = self.albumData.comment;
        self.albumComment.textColor = [UIColor piwigoColorText];
    }
    self.albumComment.font = [UIFont piwigoFontSmall];
    self.albumComment.font = [self.albumComment.font fontWithSize:[UIFont fontSizeForLabel:self.albumComment nberLines:3]];

    // Number of images and sub-albums
    self.numberOfImages.font = [UIFont piwigoFontTiny];
    self.numberOfImages.textColor = [UIColor piwigoColorText];
    NSNumberFormatter *numberFormatter = [[NSNumberFormatter alloc] init];
    [numberFormatter setNumberStyle:NSNumberFormatterDecimalStyle];
    if (self.albumData.numberOfSubCategories == 0) {
        // There are no sub-albums
        NSString *nberImages = [numberFormatter stringFromNumber:[NSNumber numberWithInteger:self.albumData.numberOfImages]];
        self.numberOfImages.text = self.albumData.numberOfImages > 1 ?
            [NSString stringWithFormat:NSLocalizedString(@"severalImagesCount", @"%@ photos"), nberImages] :
            [NSString stringWithFormat:NSLocalizedString(@"singleImageCount", @"%@ photo"), nberImages];
    }
    else if (self.albumData.totalNumberOfImages == 0) {
        // There are no images but sub-albums
        NSString *nberAlbums = [numberFormatter stringFromNumber:[NSNumber numberWithInteger:self.albumData.numberOfSubCategories]];
        self.numberOfImages.text = self.albumData.numberOfSubCategories > 1 ?
            [NSString stringWithFormat:NSLocalizedString(@"severalSubAlbumsCount", @"%@ sub-albums"), nberAlbums] :
            [NSString stringWithFormat:NSLocalizedString(@"singleSubAlbumCount", @"%@ sub-album"), nberAlbums];
    }
    else {
        // There are images and sub-albums
        NSString *nberImages = [numberFormatter stringFromNumber:[NSNumber numberWithInteger:self.albumData.totalNumberOfImages]];
        NSMutableString *nberOfImages = [[NSMutableString alloc] initWithString: self.albumData.totalNumberOfImages > 1 ?
            [NSString stringWithFormat:NSLocalizedString(@"severalImagesCount", @"%@ photos"), nberImages] :
            [NSString stringWithFormat:NSLocalizedString(@"singleImageCount", @"%@ photo"), nberImages]];
        [nberOfImages appendString:@", "];
        NSString *nberAlbums = [numberFormatter stringFromNumber:[NSNumber numberWithInteger:self.albumData.numberOfSubCategories]];
        [nberOfImages appendString:self.albumData.numberOfSubCategories > 1 ?
            [NSString stringWithFormat:NSLocalizedString(@"severalSubAlbumsCount", @"%@ sub-albums"), nberAlbums] :
            [NSString stringWithFormat:NSLocalizedString(@"singleSubAlbumCount", @"%@ sub-album"), nberAlbums]];
        self.numberOfImages.text = nberOfImages;
    }
    self.numberOfImages.font = [self.numberOfImages.font fontWithSize:[UIFont fontSizeForLabel:self.numberOfImages nberLines:1]];

    // Add renaming, moving and deleting capabilities when user has admin rights
    if(NetworkVarsObjc.hasAdminRights)
    {
        // Handle
        self.handleButton.layer.cornerRadius = 7;
        self.handleButton.backgroundColor = [UIColor piwigoColorOrange];
        self.handleButton.hidden = NO;
  
        // Left => Right swipe (only if there are images in the album)
        // Disabled because it does not work reliably on the server side
//        if (self.albumData.numberOfImages > 0) {
//
//            self.leftSwipeSettings.transition = MGSwipeTransitionBorder;
//            self.leftButtons = @[[MGSwipeButton buttonWithTitle:@""
//                                                           icon:[UIImage imageNamed:@"SwipeRefresh.png"]
//                                                backgroundColor:[UIColor blueColor]
//                                                       callback:^BOOL(MGSwipeTableCell *sender) {
//                                                           [self resfreshRepresentative];
//                                                           return YES;
//                                                       }]];
//        }
    }
    
    // Display album image
    UIImage *placeHolder = [UIImage imageNamed:@"placeholder"];
    self.backgroundImage.layer.cornerRadius = 10;
    NSInteger imageSize = CGImageGetHeight(albumData.categoryImage.CGImage) * CGImageGetBytesPerRow(albumData.categoryImage.CGImage);
    
    if (albumData.categoryImage && imageSize > 0 &&
        ![albumData.categoryImage isEqual:placeHolder])
    {
        // Album thumbnail in memory
        self.backgroundImage.image = albumData.categoryImage;
    }
    else if (albumData.albumThumbnailUrl.length <= 0)
    {
        // No album thumbnail
        albumData.categoryImage = [UIImage imageNamed:@"placeholder"];
        self.backgroundImage.image = [UIImage imageNamed:@"placeholder"];
        return;
    }
    else
    {
        // Load album thumbnail
        __weak typeof(self) weakSelf = self;
        CGSize size = self.backgroundImage.bounds.size;
        CGFloat scale = fmax(1.0, self.backgroundImage.traitCollection.displayScale);
        NSURL *URL = [NSURL URLWithString:albumData.albumThumbnailUrl];
        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:URL];
        [request addValue:@"image/*" forHTTPHeaderField:@"Accept"];
        [self.backgroundImage setImageWithURLRequest:request
                                    placeholderImage:[UIImage imageNamed:@"placeholder"]
                                             success:^(NSURLRequest *request, NSHTTPURLResponse *response, UIImage *image) {
            dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^(void){
                // Process saliency
                UIImage *croppedImage;
                if (@available(iOS 13.0, *)) {
                    croppedImage = [image processSaliency];
                    if (croppedImage == nil) { croppedImage = image; }
                } else {
                    // Fallback on earlier versions
                    croppedImage = image;
                }

                // Reduce size?
                CGSize imageSize = croppedImage.size;
                if (fmax(imageSize.width, imageSize.height) > fmax(size.width, size.height) * scale) {
                    UIImage *albumImage = [ImageUtilities downsampleWithImage:croppedImage to:size scale:scale];
                    dispatch_async(dispatch_get_main_queue(), ^{
                        albumData.categoryImage = albumImage;
                        weakSelf.backgroundImage.image = albumImage;
                    });
                } else {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        albumData.categoryImage = croppedImage;
                        weakSelf.backgroundImage.image = croppedImage;
                    });
                }
            });
         }
         failure:^(NSURLRequest *request, NSHTTPURLResponse *response, NSError *error) {
#if defined(DEBUG)
             NSLog(@"setupWithAlbumData — Fail to get album bg image for album at %@", albumData.albumThumbnailUrl);
#endif
         }];
    }
}


-(void)prepareForReuse
{
    [super prepareForReuse];
    
    [self.backgroundImage cancelImageDownloadTask];
    self.backgroundImage.image = [UIImage imageNamed:@"placeholder"];
    
    self.albumName.text = @"";
    self.numberOfImages.text = @"";
}

-(void)setFrame:(CGRect)frame
{
    frame.size.height -= 8.0;
    [super setFrame:frame];
}

@end

File: piwigo/Album/Cells/AlbumTableViewCell.swift
//
//  AlbumTableViewCell.swift
//  piwigo
//
//  Created by Spencer Baker on 1/24/15.
//  Copyright (c) 2015 bakercrew. All rights reserved.
//
//  Converted to Swift 5.4 by Eddy Lelièvre-Berna on 16/01/2022
//

import UIKit
import piwigoKit

class AlbumTableViewCell: MGSwipeTableCell {
    
    @IBOutlet weak var backgroundImage: UIImageView!
    @IBOutlet weak var topCut: UIButton!
    @IBOutlet weak var bottomCut: UIButton!
    @IBOutlet weak var albumName: UILabel!
    @IBOutlet weak var albumComment: UILabel!
    @IBOutlet weak var numberOfImages: UILabel!
    @IBOutlet weak var handleButton: UIButton!
    @IBOutlet weak var recentBckg: UIImageView!
    @IBOutlet weak var recentImage: UIImageView!
    
    private var download: ImageDownload?

    func config(withAlbumData albumData: Album?) {
        // General settings
        backgroundColor = UIColor.piwigoColorBackground()
        contentView.backgroundColor = UIColor.piwigoColorCellBackground()
        selectionStyle = UITableViewCell.SelectionStyle.none
        topCut.backgroundColor = UIColor.piwigoColorBackground()
        bottomCut.backgroundColor = UIColor.piwigoColorBackground()
        recentBckg.tintColor = UIColor(white: 0, alpha: 0.3)
        recentImage.tintColor = UIColor.white

        // Album name (Piwigo orange colour)
        var text = albumData?.name ?? "—?—"
        if albumName.text != text {
            albumName.text = text
        }
        var fontSize = fontSizeFor(label: albumName, nberLines: 2)
        var font = UIFont.systemFont(ofSize: fontSize)
        if albumName.font != font {
            albumName.font = font
        }
        
        // Album description (colour depends on text content)
        let desc = getDescription(fromAlbumData: albumData)
        if albumComment.attributedText != desc {
            albumComment.attributedText = desc
        }
        fontSize = UIFont.fontSizeFor(label: albumComment, nberLines: 3)
        font = UIFont.systemFont(ofSize: fontSize)
        if albumComment.font != font {
            albumComment.font = font
        }

        // Number of images and sub-albums
        text = getNberOfImages(fromAlbumData: albumData)
        if numberOfImages.text != text {
            numberOfImages.text = text
        }
        let color = UIColor.piwigoColorText()
        if numberOfImages.textColor != color {
            numberOfImages.textColor = color
        }
        font = UIFont.systemFont(ofSize: 10, weight: .light)
        if numberOfImages.font != font {
            numberOfImages.font = font
        }

        // Add renaming, moving and deleting capabilities when user has admin rights
        if albumData != nil, handleButton.isHidden == NetworkVars.hasAdminRights {
            handleButton.isHidden = !NetworkVars.hasAdminRights
        }

        // Display recent icon when images have been uploaded recently
        DispatchQueue.global(qos: .userInteractive).async {
            self.showHideIsRecent(albumData: albumData)
        }
        
        // Display album image
        let placeHolder = UIImage(named: "placeholder")!

        // Can we add a representative if needed?
        if albumData?.thumbnailUrl == nil || albumData?.thumbnailId == Int64.zero,
           let images = albumData?.images, let firstImage = images.first {
            // Set representative (case where images were uploaded recently)
            albumData?.thumbnailId = firstImage.pwgID
            let thumnailSize = pwgImageSize(rawValue: AlbumVars.shared.defaultAlbumThumbnailSize) ?? .medium
            albumData?.thumbnailUrl = ImageUtilities.getURL(firstImage, ofMinSize: thumnailSize) as NSURL?
        }
        
        // Do we have a representative?
        guard let thumbUrl = albumData?.thumbnailUrl,
              let thumbID = albumData?.thumbnailId,
              let serverID = albumData?.user?.server?.uuid else {
            // No album thumbnail URL
            if backgroundImage.image != placeHolder {
                backgroundImage.image = placeHolder
            }
            return
        }

        // Retrieve image from cache or download it
        let thumbSize = pwgImageSize(rawValue: AlbumVars.shared.defaultAlbumThumbnailSize) ?? .medium
        download = ImageDownload(imageID: thumbID, ofSize: thumbSize, atURL: thumbUrl as URL,
                                 fromServer: serverID, placeHolder: placeHolder) { cachedImage in
            DispatchQueue.main.async {
                self.configImage(cachedImage)
            }
        } failure: { _ in
            DispatchQueue.main.async {
                // No album thumbnail URL
                if self.backgroundImage.image != placeHolder {
                    self.backgroundImage.image = placeHolder
                }
            }
        }
        download?.getImage()
    }
    
    private func getDescription(fromAlbumData albumData: Album?) -> NSAttributedString {
        var desc = NSMutableAttributedString()
        // Any provided description?
        if let description = albumData?.comment, description.string.isEmpty == false {
            desc = NSMutableAttributedString(attributedString: description)
            let wholeRange = NSRange(location: 0, length: desc.string.count)
            let style = NSMutableParagraphStyle()
            style.alignment = NSTextAlignment.center
            let attributes = [
                NSAttributedString.Key.foregroundColor: UIColor.piwigoColorText(),
                NSAttributedString.Key.font: UIFont.systemFont(ofSize: 13, weight: .light),
                NSAttributedString.Key.paragraphStyle: style
            ]
            desc.addAttributes(attributes, range: wholeRange)
        }
        else if NetworkVars.hasAdminRights {
            let noDesc = NSLocalizedString("createNewAlbumDescription_noDescription", comment: "no description")
            desc = NSMutableAttributedString(string: noDesc)
            let wholeRange = NSRange(location: 0, length: desc.string.count)
            let style = NSMutableParagraphStyle()
            style.alignment = NSTextAlignment.center
            let attributes = [
                NSAttributedString.Key.foregroundColor: UIColor.piwigoColorRightLabel(),
                NSAttributedString.Key.font: UIFont.systemFont(ofSize: 13, weight: .light),
                NSAttributedString.Key.paragraphStyle: style
            ]
            desc.addAttributes(attributes, range: wholeRange)
        }
        return desc
    }
    
    private func getNberOfImages(fromAlbumData albumData: Album?) -> String {
        // Constants
        let numberFormatter = NumberFormatter()
        numberFormatter.numberStyle = NumberFormatter.Style.decimal
        let singleImage = NSLocalizedString("severalImagesCount", comment: "%@ photos")
        let severalImages = NSLocalizedString("singleImageCount", comment: "%@ photo")
        let singleSubAlbum = NSLocalizedString("singleSubAlbumCount", comment: "%@ sub-album")
        let severalSubAlbums = NSLocalizedString("severalSubAlbumsCount", comment: "%@ sub-albums")
        // Determine string
        var text = ""
        if albumData?.nbSubAlbums ?? 0 == 0 {
            // There are no sub-albums
            let nberImages = numberFormatter.string(from: NSNumber(value: albumData?.nbImages ?? 0))
            text = (albumData?.nbImages ?? 0 > 1)
                ? String.localizedStringWithFormat(singleImage, nberImages ?? "")
                : String.localizedStringWithFormat(severalImages, nberImages ?? "")
        }
        else if albumData?.totalNbImages ?? 0 == 0 {
            // There are no images but sub-albums
            let nberAlbums = numberFormatter.string(from: NSNumber(value: albumData?.nbSubAlbums ?? 0))
            text = (albumData?.nbSubAlbums ?? 0 > 1)
                ? String.localizedStringWithFormat(severalSubAlbums, nberAlbums ?? "")
                : String.localizedStringWithFormat(singleSubAlbum, nberAlbums ?? "")
        }
        else {
            // There are images and sub-albums
            let nberImages = numberFormatter.string(from: NSNumber(value: albumData?.totalNbImages ?? 0))
            text = (albumData?.totalNbImages ?? 0 > 1)
                ? String.localizedStringWithFormat(severalImages, nberImages ?? "")
                : String.localizedStringWithFormat(singleImage, nberImages ?? "")
            text += ", "
            let nberAlbums = numberFormatter.string(from: NSNumber(value: albumData?.nbSubAlbums ?? 0))
            text += (albumData?.nbSubAlbums ?? 0 > 1)
                ? String.localizedStringWithFormat(severalSubAlbums, nberAlbums ?? "")
                : String.localizedStringWithFormat(singleSubAlbum, nberAlbums ?? "")
        }
        return text
    }
    
    private func fontSizeFor(label: UILabel?, nberLines: Int) -> CGFloat {
        // Check label is not nil
        guard let label = label else { return 17.0 }
        let font = UIFont.systemFont(ofSize: 17)
        
        // Check that we can adjust the font
        if (label.adjustsFontSizeToFitWidth == false) ||
            (label.minimumScaleFactor >= 1.0) {
            // Font adjustment is disabled
            return font.pointSize
        }

        // Should we scale the font?
        var unadjustedWidth: CGFloat = 1.0
        if let text = label.text {
            unadjustedWidth = text.size(withAttributes: [NSAttributedString.Key.font: font]).width
        }
        let width: CGFloat = label.frame.size.width
        let height: CGFloat = unadjustedWidth / CGFloat(nberLines)
        var scaleFactor: CGFloat = width / height
        if scaleFactor >= 1.0 {
            // The text already fits at full font size
            return font.pointSize
        }

        // Respect minimumScaleFactor
        scaleFactor = fmax(scaleFactor, label.minimumScaleFactor)
        let newFontSize: CGFloat = font.pointSize * scaleFactor

        // Uncomment this if you insist on integer font sizes
        //newFontSize = floor(newFontSize);

        return newFontSize
    }

    private func showHideIsRecent(albumData: Album?) {
        guard let dateLast = albumData?.dateLast else { return }
        let timeSinceLastUpload: TimeInterval = dateLast.timeIntervalSinceNow
        var indexOfPeriod: Int = AlbumVars.shared.recentPeriodIndex
        indexOfPeriod = min(indexOfPeriod, AlbumVars.shared.recentPeriodList.count - 1)
        indexOfPeriod = max(0, indexOfPeriod)
        let periodInDays: Int = AlbumVars.shared.recentPeriodList[indexOfPeriod]
        let isRecent = timeSinceLastUpload > TimeInterval(-24*3600*periodInDays)
        DispatchQueue.main.async {
            if self.recentBckg.isHidden == isRecent {
                self.recentBckg.isHidden = !isRecent
                self.recentImage.isHidden = !isRecent
            }
        }
    }
    
    private func configImage(_ image: UIImage) {
        // Process image in the background
        DispatchQueue.global(qos: .userInitiated).async {
            // Process saliency
            var finalImage:UIImage = image
            if #available(iOS 13.0, *) {
                if let croppedImage = image.processSaliency() {
                    finalImage = croppedImage
                }
            }

            // Set image
            DispatchQueue.main.async {
                self.backgroundImage.image = finalImage
            }
        }
    }
    
    override func prepareForReuse() {
        super.prepareForReuse()
        
        download?.task?.cancel()
        download = nil
    }
}
